gremlin-dotnet/src/Gremlin.Net/Process/Traversal/DefaultTraversal.cs

```diff
@@ -86,9 +87,61 @@ public void Reset()
         }
 
         /// <inheritdoc />
-        public E Current => (E)TraverserEnumerator.Current?.Object;
+        public E Current => GetCurrent<E>();
 
-        object IEnumerator.Current => Current;
+        object IEnumerator.Current => GetCurrent();
+
+        private TReturn GetCurrent<TReturn>()
+        {
+            var value = GetCurrent();
+            var returnType = typeof(TReturn);
+            if (value == null || value.GetType() == returnType)
+            {
+                // Avoid evaluating type comparisons
+                return (TReturn) value;
+            }
+            return (TReturn) GetValue(returnType, value);
+        }
+
+        private object GetCurrent()
+        {
+            // Use dynamic to object to prevent runtime dynamic conversion evaluation
+            return TraverserEnumerator.Current?.Object;
+        }
+
+        /// <summary>
+        /// Gets the value, converting to the expected type when necessary and supported. 
+        /// </summary>
+        private static object GetValue(Type type, object value)
+        {
+            var genericType = type.GetTypeInfo().IsGenericType
+                ? type.GetTypeInfo().GetGenericTypeDefinition()
+                : null;
+            if (value is IDictionary dictValue && genericType == typeof(IDictionary<,>))
+            {
+                var keyType = type.GenericTypeArguments[0];
+                var valueType = type.GenericTypeArguments[1];
+                var mapType = typeof(Dictionary<,>).MakeGenericType(keyType, valueType);
+                var result = (IDictionary) Activator.CreateInstance(mapType);
+                foreach (DictionaryEntry kv in dictValue)
+                {
+                    result.Add(GetValue(keyType, kv.Key), GetValue(valueType, kv.Value));
+                }
+                return result;
+            }
+            if (value is IEnumerable enumerableValue && genericType == typeof(IList<>))
```

Comment:
Can't we change this to also allow `List<>`  as the `genericType` instead of `IList<>`?

Reply:
I think continue using the interfaces as the return data type allow us to hide implementation details and its an accurate representation of the Java types (`Map<,>` and `List<>` are interfaces in java).

Reply:
I don't want to change the return types for the traversal API, but a user can specify the type for some steps directly like you did in the added tests: `ValueMap<IList<int>>`. When `ValueMap<List<int>>` is used instead then it won't work right now.

Reply:
Its always easier to go from an strict API to a wider range of representations in the future than the other way around.

Reply:
Ok, then lets leave it as it is in this PR. We can still change it in case it leads to problems for users.